### ISSUE TITLE
Assertion failure when trying to connect to the same master more than once

### ISSUE THREAD

**Issue by locker** [tarantool/tarantool#3610]
Branch 1.9 commit a9695b98b6a6fa5e8c130577f274f97819d30bf4 (1.10 and 2.0 are affected as well)

How to reproduce:
1. Start a tarantool instance and type `box.cfg{listen=44441}`
2. Start another tarantool instance and type `box.cfg{replication={44441, 44441}}`

Expected an error message, but got a crash instead:
```
Tarantool 1.9.1-60-ga9695b98
type 'help' for interactive help
tarantool> box.cfg{replication={44441, 44441}}
2018-08-03 19:57:11.935 [3579] main/101/interactive C> Tarantool 1.9.1-60-ga9695b98
2018-08-03 19:57:11.935 [3579] main/101/interactive C> log level 5
2018-08-03 19:57:11.935 [3579] main/101/interactive I> mapping 268435456 bytes for memtx tuple arena...
2018-08-03 19:57:11.935 [3579] main/101/interactive I> mapping 134217728 bytes for vinyl tuple arena...
2018-08-03 19:57:11.942 [3579] main/104/applier/ I> remote master is 1.9.1 at 0.0.0.0:44441

2018-08-03 19:57:11.942 [3579] main/105/applier/ I> remote master is 1.9.1 at 0.0.0.0:44441

tarantool: /home/vlad/src/tarantool/src/box/replication.cc:150: void replica_delete(replica*): Assertion `replica_is_orphan(replica)' failed.
Aborted
```

**Comment by locker** [comment]
Another assertion of the kind:
```
Tarantool 1.9.1-60-ga9695b98
type 'help' for interactive help
tarantool> box.cfg{replication={44441, 44441}, replication_connect_quorum=0, replication_connect_timeout=1}
2018-08-03 20:27:23.527 [13197] main/101/interactive C> Tarantool 1.9.1-60-ga9695b98
2018-08-03 20:27:23.527 [13197] main/101/interactive C> log level 5
2018-08-03 20:27:23.527 [13197] main/101/interactive I> mapping 268435456 bytes for memtx tuple arena...
2018-08-03 20:27:23.528 [13197] main/101/interactive I> mapping 134217728 bytes for vinyl tuple arena...
2018-08-03 20:27:23.537 [13197] main/101/interactive I> recovery start
2018-08-03 20:27:23.537 [13197] main/101/interactive I> recovering from `./00000000000000000002.snap'
2018-08-03 20:27:23.569 [13197] main/105/applier/ I> can't connect to master
2018-08-03 20:27:23.569 [13197] main/105/applier/ coio.cc:104 !> SystemError connect, called on fd 14, aka [::1]:54808: Connection refused
2018-08-03 20:27:23.569 [13197] main/105/applier/ I> will retry every 1.00 second
2018-08-03 20:27:23.569 [13197] main/106/applier/ I> can't connect to master
2018-08-03 20:27:23.569 [13197] main/106/applier/ coio.cc:104 !> SystemError connect, called on fd 14, aka [::1]:54810: Connection refused
2018-08-03 20:27:23.569 [13197] main/106/applier/ I> will retry every 1.00 second
2018-08-03 20:27:24.569 [13197] main/101/interactive C> failed to connect to 2 out of 2 replicas
2018-08-03 20:27:24.569 [13197] main/101/interactive I> ready to accept requests
2018-08-03 20:27:24.570 [13197] main/109/checkpoint_daemon I> started
2018-08-03 20:27:24.570 [13197] main/109/checkpoint_daemon I> scheduled the next snapshot at Fri Aug  3 22:10:08 2018
---
...

tarantool> 2018-08-03 20:28:38.680 [13197] main/107/applier/ I> remote master is 1.9.1 at 0.0.0.0:44441

2018-08-03 20:28:38.681 [13197] main/108/applier/ I> remote master is 1.9.1 at 0.0.0.0:44441

2018-08-03 20:28:38.681 [13197] main/108/applier/ replication.cc:262 E> duplicate connection to the same replica: instance uuid 19806873-9809-4ce8-9462-3f4f34a0834a, addr1 44441, addr2 44441
tarantool: /home/vlad/src/tarantool/src/box/replication.cc:238: void replica_on_applier_sync(replica*): Assertion `replica->applier_sync_state == APPLIER_CONNECTED' failed.
Aborted
```
How to reproduce:
1. Start a tarantool instance and configure replication privilege for the guest user:
```
box.cfg{listen=44441}
box.schema.user.grant('guest', 'replication')
```
2. Bootstrap another tarantool instance from the instance started at step 1:
```
box.cfg{replication=44441}
```
3. Stop both instances.
4. Start the slave instance with the following configuration parameters:
```
box.cfg{replication_connect_quorum=0,
        replication_connect_timeout=1,
        replication={44441,44441}}
```
5. Start the master instance:
```
box.cfg{listen=44441}
```
6. The slave instance will crash.

**Comment by locker** [comment]
Fixed by 03a9bb1a25a99e8545e03a2fc995027644e46ddc
